### DISCUSSION TITLE
Extensions field in records for third-party usage

### DISCUSSION THREAD

**Discussion by mary-ext** [bluesky-social/atproto#1889] [Protocol (atproto)]
**Is your feature request related to a problem? Please describe.**

There is friction in third-party clients wanting to add quality-of-life functionality that requires adding additional/custom public metadata to something like profiles, where the records aren't being exposed as is to app.bsky.* endpoints (unlike posts, which does have them)

**Describe the solution you'd like**

An `extensions` field added within said record types, this is an object where basically anything goes, and is up for third-party clients to do what they want with it (can be subject to limits). This field can then be forwarded as-is to any app.bsky.* queries, without the need for said queries to expose the actual record itself.

**Describe alternatives you've considered**

Right now, clients has to retrieve two different subsets of profile informations, say, `getProfile` to get follow counts, and the actual record itself for these sorts of extensions. It's doable, but it's also a hassle to do, and takes up precious network usage that could be used towards other things (even if they're small responses)

**Additional context**

It would be nice if we can get something as basic as public profile timezones right now.

![image](https://github.com/bluesky-social/atproto/assets/148872143/ad6939e2-3355-47b8-9c93-22504024bf8d)

**Comment by bnewbold** [comment]
This is a great idea, thanks for the suggestion! I've transferred this to a longer-form "discussion".

We have had a bit of talk internally about how to better support third party extensions to record schemas. One proposal is basically exactly what you propose: a special dedicated field (probably prefixed with `$` to minimize name collisions, and make it clear this is an atproto-level design pattern), where folks can stash extra metadata. This could be "always" an object, and have first-level field names which are NSIDs, which would prevent collisions and even (in theory, eventually) allow for Lexicon specification and validation.

An alternative is to lean more heavily in to "paired records" in different Lexicon namespaces, but same `rkey` for easier lookup. The threadgate mechanism for controlling post replies uses this pattern (though that hasn't been documented or even shipped yet).

Some open questions:

- where these extension fields can exist: only at top-level? Anywhere a `$type` would be required to disambiguate the object type? Eg, if you wanted to add per-image metadata to posts
- does this mean we would change expectations and state that folks should *not* ever stash metadata in random other fields? or would both extension/evolution mechanisms be ok?
- do we want to encourage this as a full-featured extension pattern, or is it a "last resort" / "relief valve" sort of thing?

**Reply by mary-ext** [reply]
1. I could see per-image custom metadata to be a necessity, say for indicating that this one image is a speciality that points to another resource, but I'd imagine having it be this open could spiral out of control real fast. Top-level should be fine for now, but for an `object` lexicon type perhaps it might be worth having an `extensible` field that determines this?
2. I think we definitely should, although these custom metadata doesn't appear in any official means, these fields can be confusing when they do appear especially when you don't always have the documentation besides you all the time.
3. I'd say last resort, but it sure is tempting :sweat_smile:

**Comment by DavidBuchanan314** [comment]
Another potential solution is just to include the full record data in `getProfile`, in a new field. For example, `getPostThread` responses include a `record` field, which makes it easy to extract any extended information from them. This does mean some information is duplicated, but it doesn't seem so bad overall.

**Reply by bnewbold** [reply]
(necrothreading, but relevant)

yes! I think including the original record verbatim is a pattern we definitely want to encourage. have been meaning to write a Lexicon-authoring style guide and just haven't gotten to it so far.

**Comment by mikestaub** [comment]
I would be cautious to implement something custom here, as this is a generic problem that JSON schema and OpenAPI 3 have solved with $ref. How hard would it be to allow Lexicons to 'inherit and extend' other Lexicons?

**Comment by makoConstruct** [comment]
If there had been a good approach to extension from the start, lexicons never would have needed to be mutable (anyone would be able to effectively extend a type by defining a new type (with the new fields) that inherits it), and then you could have used content-addressing instead of all the reverse DNS stuff, lexicon would've been unambiguous, fully self-verifying, and to some extent more decentralized.

So this discussion is important.

In my own travels, I've been considering an approach where we structure our objects as they would be structured in any programming language that supports inheritance. I'm now wondering if atproto might be able to do this too, if there were an `$extensions` field, and if we followed certain conventions when we use it.

An object, under this approach, would consist of a tree (Or something other than a tree. See footnote: [1]) of `$type`d components that mirror the inheritance tree of the object's type. Each node in the tree links out to its supertype nodes in its `$extensions` field (which should probably be a list). Supertype nodes may have their own `$extensions` in turn.
This structure guarantees that any types could be combined through multiple inheritance, a client can always ignore type components it doesn't use, downcasts and upcasts (including chained upcasts) will always select the right segment of the component subtree, and extension field name collisions won't ever be an issue because every field resides in the component for which it was defined.

On the consumer side, it would just require clients to `upcast(o, t)` the received object `o` into the exact type `t` they expect, before they start to work with them. `upcast` would search for the required components and compile them into a nice simple (flat), consistent (for the given `t`) representation. If there were field name collisions between the components of the target type, `upcast` can prefix them. In this way, extensions would never interfere with clients who don't use those extensions, as long as the supertype component they expect is in there somewhere, upcast will pull it up.

If we followed this approach, we'd have a pretty decent decentralised type system! There are some complications we'd need to resolve though:

[1] I'm torn on whether the inheritance structure should be a tree, or a flattened array with (or without) duplicate components dropped, or a dag, or, most of the ambivalence is really about whether we should preserve the structure of "diamond problem" situations, ie, situations where you have a type that inherits types A and B where both A and B inherit C. A tree representation would mean there'd need to be two Cs. Sometimes this is good, if the C of the A means something slightly different than C of the B. IME, that kind of situation does occur. Other times, I think, more often, it's very bad, the Cs are just redundant, and in a decentralised protocol evolution situation, this is going to be a common situation, it happens whenever two authors extend a type C and then someone later defines a type that merges (extends both of) their extensions. If this merging of extensions happens repeatedly to a type that keeps growing longer over time, since each merge is duplicating the extension chain prior to it, it would actually be an exponential increase in size x_x
But if we could transmit the component hierarchy as a dag instead of a tree, component branches could share structure, if the repeat components are actually redundant, you only need to send one C which A and B both refer to in their extensions. While, if you're in one of those rare situations where they're not redundant, you can still send two Cs. You'd be able to have it either way depending on the situation. As I understand it, nodes in atproto _are_ dag-structured.
Maybe some clients would have difficulty with dags (eg, a rust client would have to use `Rc`s there (_or probably a custom ref type to permit sharing the object cache with different applications_))? But that's not really a tolerable limitation, dags are common in nature. (For that matter, [cycles are too](https://whtwnd.com/mako.dreamshrine.org/3lcdzw5oqh62t)! Though cycles definitely have no place in inheritance hierarchies x])

and

(2) Is it too late to switch to (or at least introduce) CIDs (_or technically paths using CIDs_) as references to lexicon types or NSIDs? I would hope that it isn't too late.

**Reply by makoConstruct** [reply]
Actually, internal references aren't really required. An alternative is having `upcast` infer omitted supertypes from the nearest ancestor.

EG, if you had a type that merges other profile types: `dreamshrine.profile extends atcs.name, fungus.profile, lignin.profile` where `fungus.profile` and `lignin.profile` also extend `atcs.name`, this could be encoded like:

```json
{
  "$type": "org.dreamshrine.profile",
  "bio": "delivering the packages.",
  "createdAt": "2026-08-23T04:12:00.000Z",
  "$ext": [
    { "$type": "org.atcs.name", "displayName": "sam porter" },
    {
      "$type": "com.fungus.profile",
      "favouriteFungus": "Hebeloma radicosum",
      "$ext": [
        { "$type": "org.atcs.name", "displayName": "ammonia giver" }
      ]
    },
    {
      "$type": "com.lignin.profile",
      "homeTree": "Pando",
      "$ext": [
        { "$type": "org.atcs.name", "displayName": "sam porter" }
      ]
    }
  ]
}
```

And all but one of the name entries become optional. If `upcast` is asked to cast to a fungus.profile, and it can't find a name field within the fungus.profile field, it will know to grab the name field of the nearest ancestor.

This is also an example of a situation where the same type takes on a distinction in meaning depending on its subtype: in the fungal world, sam might be known by a different name. You could keep this information, or drop it, in either case, `upcast` users who are looking for a fungus.profile will keep getting valid data.

I'm not completely sure it even needs to be a tree, rather than a flat array. I was under the impression this could result in some confusion where a supertype mistakes a type that follows it in the array for a part of itself, but I can't think of a situation where this wouldn't be correct behaviour, aside from one which violates pragmatics (if a subtype bizarrely extended `a: c`, followed by `b: c`, followed by `c`, if b's c was omitted, it would use the final `c` instead of the earliest, which would be wrong, however, why would you ever define a type that extends `c` at lower priority than `a` given that `a` provides a `c` so the final `c` would never be accessed).

**Comment by NoelDeMartin** [comment]
There are many ways in which this can be done, and I don't think I have a strong preference for the final solution... But here's some of my thoughts on things we should keep in mind:

## Client-to-client specificiations

Technically, the protocol already supports having this type of extensions. For example, we could just have related records pointing to the base record that extend the functionality.

However, what is important is that we have some best practices (or maybe "suggested" practices) on how this can be done, so that a new developer that doesn't know anything about ATProto knows how to go about it.

Still, I do think it would be better to have a built-in mechanism like `$extensions` if possible. If we want something that is easier to implement and make backwards compatible, we could also do something like [`schema:sameAs`](https://schema.org/sameAs) to make explicit the relation between records and their extensions.

## Extensions should be namespaced

This is something where I do have a strong preference. The nature of ATProto and Lexicons means that multiple people may want to create similar extensions without realizing. For the same reasons Lexicons are namespaced, extensions should be as well.

It may be a very simplistic view, but one way to see this is like adding "traits" or "behaviours" to existing Lexicons. For example, as mentioned in the initial description for this discussion we could have a trait like `my.3rd.party.HasTimezones`.

This also allows applications to ignore any extension that is unknown, and helps developers understand that it's ok to add their own data. Tim Berners-Lee gave a talk about a similar problem with Linked Data, called [Linked Data is Like a Bag of Chips](https://www.w3.org/DesignIssues/BagOfChips.html).
